### DISCUSSION TITLE
Pricing and/or other considerations regarding using Cloud Run for simulating Cloud Functions?

### DISCUSSION THREAD

**Discussion by jakub-trzebiatowski** [GoogleCloudPlatform/functions-framework-dart#309] [Q&A] [answered]
Thank you for creating this project, it looks very promising.

From a Google Cloud customer perspective, using a client-provided framework like this one to achieve a functionality otherwise provided by a dedicated service _may_ seem like a "hack" or a "workaround" with potential gotchas.

Are you aware of any pricing, performance, operational, usability-related, or any other gotchas related with using Cloud Run to "simulate" Cloud Functions functionality?

**Comment by kevmoo** [comment, marked as answer]
Hey @cubuspl42 – moved this to discussion. Since this is more of a discussion/Q&A

Cloud Functions are easier as a developer: you only have to manage the function definition itself. You don't have to worry about a Docker file, etc.

Now, we've done a LOT of work here to make the dev experience pretty close.

As far as pricing, performance, etc – I'm not sure. Your mileage may vary. I'm hoping they'd be close.

CC @subfuzion or @grant who may have other thoughts here.

**Reply by jakub-trzebiatowski** [reply]
Thank you for your answer. Indeed, I can see that there was a lot of work put in this project and I appreciate it! ❤️ 

 Unfortunately, together with my team we rejected (at this point) functions-framework-dart as a solution ot our problem. We chose to implement our solution using Cloud Functions / node.js / TypeScript, although this meant some level of code duplication.

We really _wanted_ to use functions-framework-dart, as a perspective of having our mobile app and its backend service implemented in a single programming language (Dart) sounded very tempting.

The reason was mostly missing docs (and, possibly, features; but I'm still not sure about this one). We still don't know how to quickly build & deploy a simple hello world project and our time resources are limited.

I'm saying this to you only in hope that such information is useful.

We will be watching functions-framework-dart development closely, hoping to use it at some point in the future!

**Reply by kevmoo** [reply]
I'm happy to chat more about how to get you started. You can email me `<github-alias>@google.com`

**Comment by AdamBridges** [comment]
I had a similar thought and here are the limitations for the **free tier** that I have discovered when it comes to using Cloud Run to simulate Cloud Functions:

**Cloud Functions:**
- 2 million invocations per month (includes both background and HTTP invocations)
- 400,000 GB-seconds, 200,000 GHz-seconds of compute time
- 5 GB network egress per month

**Cloud Run:**
- 2 million requests per month
- 360,000 GB-seconds of memory, 180,000 vCPU-seconds of compute time
- 1 GB network egress from North America per month

[Google Cloud Free Program](https://cloud.google.com/free/docs/gcp-free-tier#always-free-usage-limits)

Financially speaking, if you do not expect anything or anyone from outside of North America to download huge amounts of data from your Cloud Run containers/instances in any given month, then I think you'll probably be fine. Otherwise, you will want to learn JavaScript or TypeScript and use Cloud Functions as intended to minimise costs. I've started to and there's plenty of documentation and guides to reference, so it's not too bad: [Getting Started with Cloud Functions for Firebase](https://www.youtube.com/watch?v=DYfP-UIKxH0&list=PLl-K7zZEsYLkPZHe41m4jfAxUi0JjLgSM&index=1)

Note that it is probably possible to have the `functions_framework` package just convert Dart to JavaScript and have complete integration for Cloud Functions but who has the time to do the work. :(

**Reply by AdamBridges** [reply]
Looks like someone has done a lot of the work already: 

https://pub.dev/packages/firebase_functions_interop 
https://github.com/pulyaevskiy/firebase-functions-interop

**Reply by jakub-trzebiatowski** [reply]
It's very interesting that Cloud Run (as opposed to Firebase Functions) is somewhat tied to NA region. I had no idea.
